Add release checklist (#1316)
paritytech/polkadot: RELEASE.md

# Release Checklist

The following checks should be completed before releasing a new version of the Polkadot/Kusama/Westend runtime or client:

- [ ] Verify new migrations complete successfully, and the runtime state is correctly updated.
- [ ] Verify previously completed migrations are removed. (`on_runtime_upgrade`)

> - [ ] Verify pallet and extrinsic ordering has stayed the same. (check metadata)

Reviewer: ```suggestion
- [ ] Verify pallet and extrinsic ordering has stayed the same (check metadata). Bump `transaction_version` if not.
```

- [ ] Verify new extrinsics have been correctly whitelisted/blacklisted for proxy filters.

> - [ ] Check that the new client releases have run on the network without issue for 24 hours.

Reviewer 2: I would expand this a bit:
- Which network(s)?
- In what role should the node run?
- Is this in addition to having any necessary PRs run on a validator node as part of a burn-in process prior to being merged with `master`?

- [ ] Verify benchmarks have been updated for any modified runtime logic.

> - [ ] Verify Polkadot JS Apps is up to date with the latest runtime changes.

Reviewer: A lot of other projects depend on API. But maybe Apps implies API?
```suggestion
- [ ] Verify Polkadot JS API and Apps are up to date with the latest runtime changes.
```

Author: Yeah, API is better here than Apps.

# Release Checklist

> The following checks should be completed before releasing a new version of the
> Polkadot/Kusama/Westend runtime or client:

Reviewer 2: ```suggestion
The following checks should be completed before publishing a new release of the
Polkadot/Kusama/Westend runtime or client:
```

Just clarifying the language - It has caused confusion before since we have a draft release in paritytech/polkadot/releases. This checklist refers to the checks we should make before we *publish* the release.

### All Releases

> - [ ] Check that the new client releases have [run on the network](#burn-in) without issue for 24

Reviewer 2: ```suggestion
- [ ] Check that the new client versions have [run on the network](#burn-in) without issue for 24
```

  hours.
- [ ] Add [build artifacts](#build-artifacts) to the release.

### Runtime Releases

- [ ] List any [native runtime](#native-runtimes) versions associated with the release.
- [ ] Increment [`spec_version`](#spec-version) for any native runtimes.
- [ ] Verify [new migrations](#new-migrations) complete successfully, and the runtime state is
  correctly updated.
- [ ] Verify previously [completed migrations](#old-migrations-removed) are removed.
- [ ] Verify pallet and [extrinsic ordering](#extrinsic-ordering) has stayed the same. Bump
  `transaction_version` if not.
- [ ] Verify new extrinsics have been correctly whitelisted/blacklisted for
  [proxy filters](#proxy-filtering).
- [ ] Verify [benchmarks](#benchmarks) have been updated for any modified runtime logic.
- [ ] Verify [Polkadot JS API](#polkadot-js) are up to date with the latest runtime changes.

## Notes

### Burn In

Ensure that Parity DevOps has run the new release on Westend, Kusama, and Polkadot validators for
at least 24 hours prior to publishing the release.

### Build Artifacts

Add any necessary assets to the release. They should include:

- Linux binary

> - GPG signature

Reviewer 2: ```suggestion
- GPG signature of the Linux binary
```

- SHA256 of binary
- Source code
- Wasm binaries of any runtimes

> ### Native Runtimes
> 
> The release notes should list which native runtimes and their versions are included.

Reviewer 2: ```suggestion
### Release notes

The release notes should list:

- The priority of the release (i.e., how quickly users should upgrade)
- Which native runtimes and their versions are included
- The proposal hashes of the runtimes as built with [srtool](https://gitlab.com/chevdor/srtool)

The release notes may also list:

- Free text at the beginning of the notes mentioning anything important regarding this release
- Notable changes (those labelled with B[1-9]-* labels) separated into sections
- 
```

Expanded this section a bit to codify the layout of the release notes in general.

(e.g. runtime v12 corresponds to v0.8.12, even if the current runtime is not v11).

### New Migrations

Ensure that any migrations that are required due to storage or logic changes are included in the
`on_runtime_upgrade` function of the appropriate pallets.

### Old Migrations Removed

Any previous `on_runtime_upgrade` functions from old upgrades must be removed to prevent them from
executing a second time.

### Extrinsic Ordering

Offline signing libraries depend on a consistent ordering of call indices and functions. Compare
the metadata of the current and new runtimes and ensure that the `module index, call index` tuples
map to the same set of functions. In case of a breaking change, increase `transaction_version`.

Note: Adding new functions to the runtime does not constitute a breaking change as long as they are
added to the end of a pallet (i.e., does not break any other call index).

### Proxy Filtering

The runtime contains proxy filters that map proxy types to allowable calls. If the new runtime
contains any new calls, verify that the proxy filters are up to date to include them.

> ### Benchmarcks

Reviewer 2: ```suggestion
### Benchmarks
```

# Release Checklist

The following checks should be completed before publishing a new release of the
Polkadot/Kusama/Westend runtime or client:

### All Releases

> - [ ] Check that the new client versions have [run on the network](#burn-in) without issue for 24

Reviewer 3: ```suggestion
- [ ] Check that the new client versions have [run on the network](#burn-in) without issue for 12
```

24 hours seems like a lot.
I think it's reasonable to deploy the burn-in on the evening, and check the next morning that nothing went wrong.

  hours.
- [ ] Add [build artifacts](#build-artifacts) to the release.

### Runtime Releases

- [ ] List any [native runtime](#native-runtimes) versions associated with the release.

> - [ ] Increment [`spec_version`](#spec-version) for any native runtimes.

Reviewer 4: ```suggestion
- [ ] Has incremented  [`spec_version`](#spec-version) for any native runtimes from any existing use on public (non-private/test) networks.
```

- [ ] Verify [new migrations](#new-migrations) complete successfully, and the runtime state is
  correctly updated.
- [ ] Verify previously [completed migrations](#old-migrations-removed) are removed.
- [ ] Verify pallet and [extrinsic ordering](#extrinsic-ordering) has stayed the same. Bump
  `transaction_version` if not.
- [ ] Verify new extrinsics have been correctly whitelisted/blacklisted for
  [proxy filters](#proxy-filtering).
- [ ] Verify [benchmarks](#benchmarks) have been updated for any modified runtime logic.
- [ ] Verify [Polkadot JS API](#polkadot-js) are up to date with the latest runtime changes.

## Notes

### Burn In

Ensure that Parity DevOps has run the new release on Westend, Kusama, and Polkadot validators for

> at least 24 hours prior to publishing the release.

Reviewer 3: ```suggestion
at least 12 hours prior to publishing the release.
```

### Build Artifacts

Add any necessary assets to the release. They should include:

- Linux binary
- GPG signature of the Linux binary
- SHA256 of binary
- Source code
- Wasm binaries of any runtimes

### Release notes

The release notes should list:

- The priority of the release (i.e., how quickly users should upgrade)
- Which native runtimes and their versions are included
- The proposal hashes of the runtimes as built with [srtool](https://gitlab.com/chevdor/srtool)

The release notes may also list:

- Free text at the beginning of the notes mentioning anything important regarding this release
- Notable changes (those labelled with B[1-9]-* labels) separated into sections

> - 

Reviewer 5: ```suggestion
```
